PR: https://github.com/odoo/odoo/pull/73992

From: a96757e19cfa3df6ac60c6f4304874a4eb7334e4
From: Csaba Tóth
Date: 2021-08-09 08:30:31

Structural Changes: 1
Total Changes: 35

[FIX] Fix journal dashboard late counter

Problem: The journal dashboard show wrong values for the late invoices.
Background details: The filter for the "Late" invoices is not in sync with the sql query in the dashboard python code. The filter searchs for the invoice_date_due field while the sql searchs for the date field.

closes odoo/odoo#73992

X-original-commit: f1e9e756b8806aca92fda4d528fd0b23842e967a
Signed-off-by: William André (wan) <wan@odoo.com>

================================= pseudo patch: =================================

--- a/addons/account/models/account_journal_dashboard.py
+++ b/addons/account/models/account_journal_dashboard.py
@@ -275,23 +275,10 @@ class account_journal(models.Model):
             self.env.cr.execute(query, query_args)
             query_results_drafts = self.env.cr.dictfetchall()
 
-            today = fields.Date.context_today(self)
-            query = '''
-                SELECT
-                    (CASE WHEN move_type IN ('out_refund', 'in_refund') THEN -1 ELSE 1 END) * amount_residual AS amount_total,
-                    currency_id AS currency,
-                    move_type,
-                    invoice_date,
-                    company_id
-                FROM account_move move
-                WHERE journal_id = %s
-                AND date <= %s
-                AND state = 'posted'
-                AND payment_state in ('not_paid', 'partial')
-                AND move_type IN ('out_invoice', 'out_refund', 'in_invoice', 'in_refund', 'out_receipt', 'in_receipt');
-            '''
-            self.env.cr.execute(query, (self.id, today))
+            (query, query_args) = self._get_late_bills_query()
+            self.env.cr.execute(query, query_args)
             late_query_results = self.env.cr.dictfetchall()
+
             curr_cache = {}
             (number_waiting, sum_waiting) = self._count_results_and_sum_amounts(query_results_to_pay, currency, curr_cache=curr_cache)
             (number_draft, sum_draft) = self._count_results_and_sum_amounts(query_results_drafts, currency, curr_cache=curr_cache)
@@ -371,6 +358,22 @@ class account_journal(models.Model):
             AND move.move_type IN ('out_invoice', 'out_refund', 'in_invoice', 'in_refund', 'out_receipt', 'in_receipt');
         ''', {'journal_id': self.id})
 
+    def _get_late_bills_query(self):
+        return """
+            SELECT
+                (CASE WHEN move_type IN ('out_refund', 'in_refund') THEN -1 ELSE 1 END) * amount_residual AS amount_total,
+                currency_id AS currency,
+                move_type,
+                invoice_date,
+                company_id
+            FROM account_move move
+            WHERE journal_id = %(journal_id)s
+            AND invoice_date_due <= %(today)s
+            AND state = 'posted'
+            AND payment_state in ('not_paid', 'partial')
+            AND move_type IN ('out_invoice', 'out_refund', 'in_invoice', 'in_refund', 'out_receipt', 'in_receipt');
+        """, {'journal_id': self.id, 'today': fields.Date.context_today(self)}
+
     def _count_results_and_sum_amounts(self, results_dict, target_currency, curr_cache=None):
         """ Loops on a query result to count the total number of invoices and sum
         their amount_total field (expressed in the given target currency).
